src/System.IO.FileSystem/src/System/IO/File.cs

```diff
@@ -725,5 +728,317 @@ public static void Decrypt(String path)
 
             throw new PlatformNotSupportedException();
         }
+
+        // UTF-8 without BOM and with error detection.
+        // Same as the default encoding for StreamWriter.
+        private static Encoding UTF8NoBOM => _UTF8NoBOM ?? (_UTF8NoBOM = new UTF8Encoding(false, true));
+
+        // If we use the path-taking constructors we will not have FileOptions.Asynchronous set and
+        // we will have asynchronous file access faked by the thread pool. We want the real thing.
+        private static StreamReader AsyncStreamReader(string path, Encoding encoding)
+        {
+            FileStream stream = new FileStream(
+                path, FileMode.Open, FileAccess.Read, FileShare.Read, DefaultBufferSize,
+                FileOptions.Asynchronous | FileOptions.SequentialScan);
+
+            return new StreamReader(stream, encoding, true);
+        }
+
+        private static StreamWriter AsyncStreamWriter(string path, bool append, Encoding encoding)
+        {
+            FileStream stream = new FileStream(
+                path, append ? FileMode.Append : FileMode.Create, FileAccess.Write, FileShare.Read, DefaultBufferSize,
+                FileOptions.Asynchronous | FileOptions.SequentialScan);
+
+            return new StreamWriter(stream, encoding);
+        }
+
+        public static Task<string> ReadAllTextAsync(string path, CancellationToken cancellationToken = default(CancellationToken))
+            => ReadAllTextAsync(path, Encoding.UTF8, cancellationToken);
+
+        public static Task<string> ReadAllTextAsync(string path, Encoding encoding, CancellationToken cancellationToken = default(CancellationToken))
+        {
+            if (path == null)
+                throw new ArgumentNullException(nameof(path));
+            if (encoding == null)
+                throw new ArgumentNullException(nameof(encoding));
+            if (path.Length == 0)
+                throw new ArgumentException(SR.Argument_EmptyPath, nameof(path));
+
+            return cancellationToken.IsCancellationRequested
+                ? Task.FromCanceled<string>(cancellationToken)
+                : InternalReadAllTextAsync(path, encoding, cancellationToken);
+        }
+
+        private static async Task<string> InternalReadAllTextAsync(string path, Encoding encoding, CancellationToken cancellationToken)
+        {
+            Debug.Assert(!string.IsNullOrEmpty(path));
+            Debug.Assert(encoding != null);
+
+            using (StreamReader sr = AsyncStreamReader(path, encoding))
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+                StringBuilder sb = new StringBuilder();
+                int bufferSize = sr.CurrentEncoding.GetMaxCharCount(DefaultBufferSize);
+                char[] buffer = new char[bufferSize];
```

Comment:
Maybe `StringBuilderCache.Acquire()` and `ArrayPool<char>.Shared.Rent()` respectively could be used here.